### ISSUE TITLE
Untilize produces garbage for TILE + HEIGHT_SHARDED input with non-tile-aligned dim

### ISSUE THREAD

**Issue by rpavlovicTT** [tenstorrent/tt-metal#39074]
## Bug Description

`ttnn.to_layout(tensor, ROW_MAJOR_LAYOUT, memory_config=interleaved_l1)` produces garbage output when the input tensor is **TILE layout + HEIGHT_SHARDED** and has a **non-tile-aligned dimension** (e.g. dim[-2] = 17).

The same conversion from TILE + INTERLEAVED works correctly (PCC = 1.0), confirming the bug is specific to the sharded untilize path.

This also affects `ttnn.slice` on sharded + tiled inputs, since slice internally calls `to_layout` to convert to ROW_MAJOR before slicing (`slice.cpp:115`).

Additionally, calling `ttnn.to_layout` without an explicit `memory_config` (which defaults to keeping the sharded config) causes a **hang** rather than an error.

## Reproduction

```python
import torch
import ttnn


def compute_pcc(a, b):
    a = a.float().flatten()
    b = b.float().flatten()
    if a.numel() != b.numel() or a.numel() <= 1:
        return float("nan")
    a_mean, b_mean = a.mean(), b.mean()
    a_std, b_std = a.std(), b.std()
    if a_std == 0 and b_std == 0:
        return 1.0
    if a_std == 0 or b_std == 0:
        return 0.0
    a_c = a - a_mean
    b_c = b - b_mean
    return ((a_c * b_c).sum() / (torch.sqrt((a_c**2).sum()) * torch.sqrt((b_c**2).sum()))).item()


def main():
    device = ttnn.open_mesh_device(ttnn.MeshShape(1, 1))

    torch_input = torch.randn(32, 32, 17, 64, dtype=torch.bfloat16)

    dram_mem = ttnn.MemoryConfig(ttnn.TensorMemoryLayout.INTERLEAVED, ttnn.BufferType.DRAM)
    l1_mem = ttnn.MemoryConfig(ttnn.TensorMemoryLayout.INTERLEAVED, ttnn.BufferType.L1)

    tt_tile = ttnn.from_torch(
        torch_input, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
        device=device, memory_config=dram_mem,
    )

    # Height sharded: 8x8 grid (64 cores), shard [512, 64]
    # Padded shape [32,32,32,64] -> 32*32*32=32768 sticks / 64 cores = 512
    hs_mem = ttnn.MemoryConfig(
        ttnn.TensorMemoryLayout.HEIGHT_SHARDED,
        ttnn.BufferType.L1,
        ttnn.ShardSpec(
            ttnn.CoreRangeSet([ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(7, 7))]),
            [512, 64],
            ttnn.ShardOrientation.ROW_MAJOR,
        ),
    )
    tt_sharded = ttnn.to_memory_config(tt_tile, hs_mem)

    # Bug: this conversion produces garbage
    tt_rm = ttnn.to_layout(tt_sharded, ttnn.ROW_MAJOR_LAYOUT, memory_config=l1_mem)

    result = ttnn.to_torch(ttnn.from_device(tt_rm))
    pcc = compute_pcc(torch_input, result)

    print(f"Shape:  {list(torch_input.shape)} (dim[-2]=17, non-tile-aligned)")
    print(f"Input:  TILE + HEIGHT_SHARDED")
    print(f"Output: ROW_MAJOR + INTERLEAVED L1")
    print(f"PCC:    {pcc:.6f} {'PASS' if pcc >= 0.99 else 'FAIL'}")

    # Control: same conversion from INTERLEAVED TILE works fine
    tt_rm_ctrl = ttnn.to_layout(tt_tile, ttnn.ROW_MAJOR_LAYOUT, memory_config=l1_mem)
    result_ctrl = ttnn.to_torch(ttnn.from_device(tt_rm_ctrl))
    pcc_ctrl = compute_pcc(torch_input, result_ctrl)
    print(f"Control (INTERLEAVED TILE -> RM): PCC={pcc_ctrl:.6f} {'PASS' if pcc_ctrl >= 0.99 else 'FAIL'}")

    ttnn.close_mesh_device(device)


if __name__ == "__main__":
    main()
```

## Output

```
Shape:  [32, 32, 17, 64] (dim[-2]=17, non-tile-aligned)
Input:  TILE + HEIGHT_SHARDED
Output: ROW_MAJOR + INTERLEAVED L1
PCC:    0.001817 FAIL
Control (INTERLEAVED TILE -> RM): PCC=1.000000 PASS
```

## Expected Behavior

PCC should be ~1.0 for both paths. The untilize result should be identical regardless of whether the source tensor is sharded or interleaved.

## Impact

- `ttnn.to_layout` (untilize) produces corrupt data for TILE + HEIGHT_SHARDED inputs with non-tile-aligned dimensions
- `ttnn.slice` on sharded + tiled tensors is broken as a consequence (it calls `to_layout` internally at `slice.cpp:115`)
- `ttnn.to_layout` without explicit `memory_config` hangs instead of erroring

## Environment

- Wormhole B0
- Main branch

**Comment by ronnie00TT** [comment]
Supported in PR #49489
